data-plane/gateway/service/src/streaming.rs

```diff
@@ -0,0 +1,994 @@
+        session_direction: SessionDirection,
+        tx_gw: GwChannelSender,
+        tx_app: AppChannelSender,
+    ) -> Streaming {
+        let (tx, rx) = mpsc::channel(128);
+        let (timer_tx, timer_rx) = mpsc::channel(128);
+        if session_direction == SessionDirection::Sender {
+            let prod = Producer {
+                buffer: ProducerBuffer::with_capacity(500),
+                next_id: 0,
+            };
+            let stream = Streaming {
+                common: Common::new(
+                    id,
+                    session_direction,
+                    SessionConfig::Streaming(session_config),
+                    tx_gw,
+                    tx_app,
+                ),
+                buffer: Arc::new(RwLock::new(Endpoint::Producer(prod))),
+                tx,
+            };
+            stream.process_message(rx, timer_rx);
+            stream
+        } else {
+            let observer = RtxTimerObserver { channel: timer_tx };
+
+            let recv = Receiver {
+                config: session_config.clone(),
+                buffer: ReceiverBuffer::default(),
+                timer_observer: Arc::new(observer),
+                rtx_map: HashMap::new(),
+                timers_map: HashMap::new(),
+            };
+
+            let stream = Streaming {
+                common: Common::new(
+                    id,
+                    session_direction,
+                    SessionConfig::Streaming(session_config),
+                    tx_gw,
+                    tx_app,
+                ),
+                buffer: Arc::new(RwLock::new(Endpoint::Receiver(recv))),
+                tx,
+            };
+
+            stream.process_message(rx, timer_rx);
+            stream
+        }
+    }
+
+    fn process_message(
+        &self,
+        mut rx: mpsc::Receiver<Result<(Message, MessageDirection), Status>>,
+        mut timer_rx: mpsc::Receiver<Result<(u32, bool), Status>>,
+    ) {
+        let buffer_clone = self.buffer.clone();
+        let session_id = self.common.id();
+        let send_gw = self.common.tx_gw();
+        let send_app = self.common.tx_app();
+        let (source, max_retries, timeout) = match self.common.session_config() {
+            SessionConfig::Streaming(streaming_configuration) => (
+                streaming_configuration.source,
+                streaming_configuration.max_retries.unwrap_or(0),
+                streaming_configuration
+                    .timeout
+                    .unwrap_or(std::time::Duration::new(0, 0)),
+            ),
+            _ => {
+                panic!("unable to parse streaming configuration");
+            }
+        };
+        let mut producer_name: Option<Agent> = None;
+        //let state = self.common.state().clone();
+        let mut timer_rx_closed = false;
+        tokio::spawn(async move {
+            debug!("starting message processing on session {}", session_id);
+            loop {
```

Comment:
We need a cancellation token or to pass the drain signal to trigger the loop stop.